src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs

```diff
@@ -363,107 +367,129 @@
+                        // .NET Core: Directory.Enumerate with empty filter returns all files in directory.
+                        // We will standardize on the .NET Core behavior.
+                        if (string.IsNullOrEmpty(filter))
+                        {
+                            filter = "*";
+                        }
 #endif
 
-                    foreach (string file in FileSystem.DirectoryEnumerateFiles(directory, filter, SearchOption.TopDirectoryOnly))
-                    {
-                        // Only include files that are below the max size limit.
-                        if (IsTargetWithinFileSizeLimit(file, _rootContext.MaxFileSizeInKilobytes))
+                        foreach (string file in FileSystem.DirectoryEnumerateFiles(directory, filter, SearchOption.TopDirectoryOnly))
                         {
-                            sortedFiles.Add(file);
-                            continue;
+                            // Only include files that are below the max size limit.
+                            if (ShouldEnqueue(file, _rootContext))
+                            {
+                                sortedFiles.Add(file);
+                                continue;
+                            }
+                            _ignoredFilesCount++;
                         }
-                        _ignoredFilesCount++;
-                    }
 
-                    foreach (string file in sortedFiles)
-                    {
-                        _fileContexts.TryAdd(
-                            _fileContextsCount,
-                            CreateContext(options,
-                                          new CachingLogger(options.Level, options.Kind),
-                                          rootContext.RuntimeErrors,
-                                          rootContext.Policy,
-                                          filePath: file)
-                        );
-
-                        await readyToHashChannel.Writer.WriteAsync(_fileContextsCount++);
+                        foreach (string file in sortedFiles)
+                        {
+                            _fileContexts.TryAdd(
+                                _fileContextsCount,
+                                CreateContext(options,
+                                              new CachingLogger(options.Level, options.Kind),
+                                              _rootContext.RuntimeErrors,
+                                              _rootContext.Policy,
+                                              filePath: file)
+                            );
+
+                            await readyToHashChannel.Writer.WriteAsync(_fileContextsCount++);
+                        }
                     }
                 }
             }
-
-            readyToHashChannel.Writer.Complete();
+            catch (Exception e)
+            {
```

Comment:
OK, so here's the source of some of the hangs that have been reported, if file enumeration raised an unhandled exception, we never closed the write channel. Oops.